File: Src/ExpectationLib/Observer.cs
using System.Text;

namespace ExpectationLib;

public class Observer: IOverseer
{
    private readonly object _lockObject = new();
    public List<Observation> Observations { get; } = new();
        
    public Observation Observe(IStimulus stimulus, IResponse response, string context = "")
    {
        lock (_lockObject)
        {
            var observation = new Observation(stimulus, response, context);
            Observations.Add(observation);
            return observation;
        }
    }

    public Observation Observe(ICircumstance circumstance, string context = null) => Observe(circumstance.Stimulus, circumstance.Response, context);

    public override string ToString()
    {
        var sb = new StringBuilder();
        var i = 0;
        foreach (var observation in Observations)
        {
            sb.Append($" #{++i}: {observation.Stimulus.Sender.Role} -> {observation.Stimulus.Receiver.Role} (Response Context: {observation.Response.Context}) (Observation Context: {observation.Context})\n");
        }
        return sb.ToString();
    }
}

File: Src/CppLibExpectation/Observer.cpp
#include "Observer.h"

namespace ExpectationLib
{
	std::shared_ptr<Observation> Observer::Observe(const std::shared_ptr<ICircumstance>& circumstance,
	                                               const std::string observationContext)
	{
		return Observe(circumstance->GetStimulus(), circumstance->GetResponse(), observationContext);
	}

	std::shared_ptr<Observation> Observer::Observe(const std::shared_ptr<IStimulus>& stimulus,
	                                               const std::shared_ptr<IResponse>& response,
	                                               const std::string& observationContext)
	{
		auto observation = std::make_shared<Observation>(stimulus, response, observationContext);
		Observations.push_back(observation);
		return observation;
	}

	std::shared_ptr<Observation> Observer::Observe(const std::shared_ptr<IStimulus>& stimulus,
		const std::string& observationContext)
	{
		const std::shared_ptr<IResponse> stimulusResponse = stimulus->GetResponse();
		return Observe(stimulus, stimulusResponse,observationContext);
	}
}
